register-machines/src/main.rs

```diff
@@ -13,66 +15,165 @@ enum Instruction {
 
 use Instruction::*;
 fn eval_program(program: &[Instruction], state: &State) -> State {
-    unimplemented!();
+    let (mut l, mut rs) = state.clone();
+    while l < program.len() {
+        match program[l] {
+            Add(i, j) => {
+                *rs.entry(i).or_insert(Godel::new()) += 1;
+                l = j; 
+            },
+            Sub(i, j, k) => {
+                let v = rs.entry(i).or_insert(Godel::new()); 
+                if *v != 0 {
+                    *v -= 1;
+                    l = j;
+                } else {
+                    l = k; 
+                }
+            },
+            Halt => { break; }
+        }
+    }
+    (l, rs)
+}
+fn encode_instruction(ins: &Instruction) -> Godel {
+    match ins {
+        Add(i, j) => 
+            encode_pair1(&Godel::from(2 * i), &Godel::from(*j)),
+        Sub(i, j, k) => 
+            encode_pair1(&Godel::from(2 * i + 1),
+            &encode_pair2(&Godel::from(*j), &Godel::from(*k))
+        ),
+        Halt => Godel::ZERO
+    }
 }
 // <<x,y>> = (2^x)*(2y+1)
-fn encode_pair1(x: u128, y: u128) -> u128 {
-    unimplemented!();
+fn encode_pair1(x: &Godel, y: &Godel) -> Godel {
+    (Godel::from(2) * y + 1) << x.to_u32_wrapping()
 }
 // <x,y> = (2^x)*(2y+1)-1
-fn encode_pair2(x: u128, y: u128) -> u128 {
-    unimplemented!();
+fn encode_pair2(x: &Godel, y: &Godel) -> Godel {
+    encode_pair1(x, y) - 1
 }
-fn encode_list_to_godel(l: &[u128]) -> u128 {
-    unimplemented!();
+fn encode_list_to_godel(l: &[Godel]) -> Godel {
+    if l.is_empty() { return Godel::ZERO; } 
+    encode_pair1(&l[0], &encode_list_to_godel(&l[1..])) 
 }
```

Comment:
This is a pretty implementation. This can be hard to spot, but what you've defined is a fold over the reversed `l`. This can be achieved far more efficiently with: `l.iter().rev().fold(Godel::ZERO, |acc, v| encode_pair1(*v, acc))`
To demonstrate the difference (using u128s) I've compiled both variants here to show you what I mean:
Recursive method:
```
example::encode_list_to_godel_recursive:
        push    r14
        push    rbx
        push    rax
        test    rsi, rsi
        je      .LBB1_1
        mov     rbx, qword ptr [rdi]
        mov     r14, qword ptr [rdi + 8]
        add     rdi, 16
        add     rsi, -1
        call    qword ptr [rip + example::encode_list_to_godel_recursive@GOTPCREL]
        mov     rcx, rdx
        mov     rdi, rbx
        mov     rsi, r14
        mov     rdx, rax
        add     rsp, 8
        pop     rbx
        pop     r14
        jmp     qword ptr [rip + example::encode_pair1@GOTPCREL]
.LBB1_1:
        xor     eax, eax
        xor     edx, edx
        add     rsp, 8
        pop     rbx
        pop     r14
        ret
```
As you can see, there's no tail-call optimisation here, and your space complexity is `O(n)`. In an ideal world, the compiler would recognise that this is a reversed fold and be happy with that, but it's not quite there yet so we need to be careful with our implementations.

The fold solution:
```
example::encode_list_to_godel_iter:
        push    r15
        push    r14
        push    rbx
        test    rsi, rsi
        je      .LBB2_1
        mov     rbx, rsi
        mov     r14, rdi
        shl     rbx, 4
        xor     eax, eax
        mov     r15, qword ptr [rip + example::encode_pair1@GOTPCREL]
        xor     ecx, ecx
.LBB2_3:
        mov     rdi, qword ptr [r14 + rbx - 16]
        mov     rsi, qword ptr [r14 + rbx - 8]
        mov     rdx, rax
        call    r15
        mov     rcx, rdx
        add     rbx, -16
        jne     .LBB2_3
        jmp     .LBB2_4
.LBB2_1:
        xor     eax, eax
        xor     ecx, ecx
.LBB2_4:
        mov     rdx, rcx
        pop     rbx
        pop     r14
        pop     r15
        ret
```
No stack frame used, so the space complexity is constant.
The interactive assembler with code is here: https://godbolt.org/z/YhEjd9176

Reply:
In fact, I find it weird that clippy isn't capable of figuring this simple case yet. I might look at implementing this as a hint for clippy as 2/3 solutions I've marked did a recursive implementation instead of a fold. I'll take a look at implementing it for clippy in the future.